src/Background/Portal.vala

```diff
@@ -0,0 +1,171 @@
+            }
+
+            results[app_id] = ApplicationState.RUNNING; //FIXME: Don't hardcode: needs implementation on the gala side
+        }
+
+        return results;
+    }
+
+    public async void notify_background (
+        ObjectPath handle,
+        string app_id,
+        string name,
+        out uint32 response,
+        out HashTable<string, Variant> results
+    ) throws DBusError, IOError {
+        uint32 _response = 2;
+        var _results = new HashTable<string, Variant> (str_hash, str_equal);
+
+        var notification_request = new NotificationRequest ();
+
+        notification_request.response.connect ((result) => {
+            switch (result) {
+                case NotificationRequest.NotifyBackgroundResult.CANCELLED:
+                    _response = 1;
+                    break;
+                case NotificationRequest.NotifyBackgroundResult.FAILED:
+                    break;
+                default:
+                    _response = 0;
+                    _results["result"] = result;
+                    break;
+            }
+
+            notify_background.callback ();
+        });
+
+        uint register_id = 0;
+        try {
+            register_id = connection.register_object<NotificationRequest> (handle, notification_request);
+        } catch (Error e) {
+            warning ("Failed to export request object: %s", e.message);
+            throw new DBusError.OBJECT_PATH_IN_USE (e.message);
+        }
+
+        notification_request.send_notification (app_id, name);
+
+        yield;
+
+        connection.unregister_object (register_id);
+        response = _response;
+        results = _results;
+    }
+
+    private enum AutostartFlags {
+        AUTOSTART_FLAGS_NONE,
+        AUTOSTART_FLAGS_DBUS_ACTIVATABLE
+    }
+
+    public bool enable_autostart (
+        string app_id,
+        bool enable,
+        string[] commandline,
+        uint32 flags
+    ) throws DBusError, IOError {
+        /* If the portal request is made by a non-flatpak application app_id will most of the time be empty */
+        if (app_id.strip () == "") {
+            /* Usually we can then asume that the first commandline arg is the app_id */
+            if (commandline[0].strip () != "") {
+                app_id = commandline[0];
+            } else {
+                return false;
+            }
+        }
+
+        var path = Path.build_filename (Environment.get_user_config_dir (), "autostart", app_id + ".desktop");
+
+        if (!enable) {
+            FileUtils.unlink (path);
+            return false;
+        }
+
+        var autostart_flags = (AutostartFlags) flags;
+
+        var key_file = new KeyFile ();
+        key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_TYPE, "Application");
+        key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_NAME, app_id);
+        key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_EXEC, flatpak_quote_argv (commandline));
```

Comment:
For display reasons, in the switchboard plug, we should query the actual application name and icon and insert then in the autostart file. the commandline goes as the Comment key.

```suggestion
        var app_info = app_id != "" ? new DesktopAppInfo (app_id + ".desktop") : null;
        var key_file = new KeyFile ();

        key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_TYPE, KeyFileDesktop.TYPE_APPLICATION);
        if (app_info != null) {
            key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_NAME, app_info.get_name ());
            key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_ICON, app_info.get_string (KeyFileDesktop.KEY_ICON));
        } else {
            key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_NAME, _("Custom Command"));
            key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_ICON, "application-default-icon");
        }

        key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_COMMENT, string.joinv (" ", commandline));
        key_file.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_EXEC, flatpak_quote_argv (commandline));
```